ipn/store/httpstore/store_http.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package httpstore contains an ipn.StateStore implementation using generic HTTP.
+//
+// Write Request
+//
+// To write state to the server, the client sends a POST request
+// with content type `application/vnd.tailscale.ipn-state+json`.
+//
+// The server is expected to return response code "204 No Content".
+// It is assumed the write request has failed in case of any other response code.
+//
+// Read Request
+//
+// To read state, the client sends a simple GET request.
+//
+// If the server has previously received a write request against the same URL,
+// it is expected to return a "200 OK" response.
+// The response should carry the content type and body as described above.
+//
+// Otherwise, the server is expected to return "204 No Content".
+//
+// It is assumed the read request has failed in case of any other response code.
+package httpstore
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"tailscale.com/ipn"
+	"tailscale.com/ipn/store/mem"
+	"tailscale.com/types/logger"
+)
+
+// httpStore is a HTTP-backed store with in-memory write-through caching.
+type httpStore struct {
+	httpClient *http.Client
+	requestURL string
+	memory     mem.Store
+}
+
+const contentType = "application/vnd.tailscale.ipn-state+json"
```

Comment:
Need confirmation from Tailscale maintainers whether it's ok to use this MIME type